### ISSUE TITLE
Convert graphic to text and visa-versa

### ISSUE THREAD

**Issue by ffd8** [basiljs/basil.js#145]
While testing the needs for a b.textMode(b.CENTER) « so one can draw text frames from center out... I realized that we don't have an easy way to place text into a normal rect = [converting a graphic to text frame](https://forums.adobe.com/thread/611553). 

Therefore I'd suggest a basic contentType switcher. Routes I could imagine going down are:

Processing + beginner style:
`b.toText(obj);`
`b.toGraphic(obj);`

`b.contentType(obj, b.CONSTANT);`
`b.TEXT_TYPE`, `b.GRAPHIC_TYPE`, `b.UNASSIGNED`

In the background, they do the same thing, ie:

`obj.contentType = ContentType.TEXT_TYPE; `
`obj = obj.getElements()[0];`

I've recalled the past two weeks how 1st timer struggle with knowing the right wording/capitalization for constants.. but they do let us pack things into single function. Started off for the `b.toText()` style.. but slowly leaning towards `b.contentType()` – since one has to read the reference/docu anyways to know about the function.

Thoughts?

**Comment by b-g** [comment]
@ffd8 Interesting! Never had that problem.

Q: Are the two type castings `Graphic > Text` and `Textframe > Graphic` the only cases?

Also I find the indesign wording confusing as my first association with "TEXT" was not a Textframe but rather the text inside a textframe.

**Comment by trych** [comment]
Could this maybe be implemented into `b.text()` and `b.image()` directly?
What we can do at the moment is this

```js
var myEllipse = b.ellipse(20, 20, 80, 80);
var myImage = b.image("img.jpg", ellipse);
```
So we might as well also allow stuff like

```js
var myTextFrame = b.text(b.LOREM, 20, 20, 80, 80);
var myImage = b.image("img.jpg", myTextFrame); // --> turns myTextFrame into graphic frame
```

and

```js
var myEllipse = b.ellipse(20, 20, 80, 80);
var myImage = b.text(b.LOREM, myEllipse); // --> places text into the ellipse
```

On your other point

> While testing the needs for a `b.textMode(b.CENTER)`

thought about this as well recently. I actually always wondered, why textFrames don't use the currently set `rectMode` by default. I would find this the most intuitive solution (yes, I know a `textFrame` is not a `rect`in InDesign terms, but it is in fact a rectangle in the end. ;) )

**Comment by fabianmoronzirfas** [comment]
what about having a new function called something like 

    b.place(filepath | String , frame);

To place an image or text document or a string into a frame (where frame can be oval, rect, polygon, textFrame etc).  

This would be more the "InDesign-Way".  

+1 on drawing textFrames from the center.

**Comment by b-g** [comment]
+1 `b.place(obj | filepath | String , frame)`

+1 drawing textFrames from the center

BTW another annoying textframe related issue for me is, that you always have to explicitly set width and height ... cooler would be to make them optional and fit the size of the frame automatically to the text e.g. `b.text("Hi!",100,100)`

**Comment by fabianmoronzirfas** [comment]
> cooler would be to make them optional and fit the size of the frame automatically to the text  

Fit it to where? The height? The width?  

It would be cool. But we have to define at least one of the width and height to have something to fit to. Or we need a default width for the textFrame. Could be the margins width. Still this would make not really sense if we have the possibility to define the location of the frame.

**Comment by b-g** [comment]
> Fit it to where? The height? The width?

Well, see the magic below :)
I guess this feature is exposed to extend-script? (never tried)

This ugly textframe ...
![fit frome to content 1](https://cloud.githubusercontent.com/assets/480224/20834891/d61304ca-b897-11e6-9ff4-a1ac4cf41792.png)

becomes a tidy one!
![fit frome to content 2](https://cloud.githubusercontent.com/assets/480224/20834892/d6134b9c-b897-11e6-9701-ef9a80b62526.png)

**Comment by ffd8** [comment]
Haha- yup I have students use this all the time, code = 

`obj.fit(FitOptions.FRAME_TO_CONTENT); `

It's great, for last stage fitting– so we could allow only placement and draw a full page textFrame, then scale it at the end? Only issues you run into, is if changing the font-size after.. or doing some b.words/characters, quickly running text-overflow... but for the ease of simply writing `b.text("string", x, y);` it's totally worth it!

**Comment by fabianmoronzirfas** [comment]
Woah. This is witchcraft! 🎩🐇 
So width and height are only applied by \n and \r in the string. Okay now I get it. 😏

**Comment by fabianmoronzirfas** [comment]
Still. We need to give a default width and height and fit at the end.

**Comment by ffd8** [comment]
@b-g, there are 3 types, `b.TEXT_TYPE, b.GRAPHIC_TYPE, b.UNASSIGNED` – Not exactly sure what the last one is.. Anyways, I would not modify the b.image/b.text to act as a placer = pretty confusing.. I'm also in favor of a `b.place();`.

Question is the order of passing variables:

`b.place(contents_obj, frame_obj);`
vs.
`b.place(frame_obj, contents_obj);`

I need to think about it more, but the 2nd one makes more sense right now.. first saying who is going to get some content (the frame to be modified/added to), then what contents go into it.. This `b.place()` function could do the content_type modifications automatically.. but it could still be useful to explicitly convert one to another = `b.toTextframe(obj);` + `b.toGraphic(obj);` +/- `b.toUnassigned(obj);`. Still useful if needing to treat it as such without actually placing anything.

**Comment by ffd8** [comment]
@fabianmoronzirfas Re: default width + height.. true, I'd just set it as b.width + b.height– then they could place the textFrame anywhere, it would be huge.. and only if they didn't explicitly give a width/height– it would then do this framefitting trim. If one creates a textFrame w/ width + height, we ignore this fitting and let them deal with it on their own (we can also make a simple `b.fitFrame(obj, [optional fitting options if applicable for same image frame]);` <- textFrames only have that one option... but images have more.. so b.fit() can require just one by default, but allow more if applicable. I think there's already another issue on fitting.

**Comment by ffd8** [comment]
@trych I totally agree that our `b.text()` should just follow the setting for `b.rectMode()`, since it is a rectangle[ular form] and would be less confusing than an own `textMode();`.

**Comment by trych** [comment]
+1 for allowing `b.text()` with 3 or 1 arguments only.

> Anyways, I would not modify the b.image/b.text to act as a placer

I don't think it's confusing at all. First tell us which image file and then tell us where to place it (rect, oval, polygon, textFrame) or alternatively tells us the coordinates.

This is btw. how `b.image()` *already works* (minus the textFrame) and my students use it without any problems.

And I think the same could and should be done for text: Tells us which string to use and then tell us where to place it: textFrame, rect, oval, poly, *path*. Otherwise, tell us some coordinates. I would find this a very logical enhancement of `b.text()` and `b.image()`.

**Comment by trych** [comment]
The two alternative workflow suggestions (to for example place text in an ellipse) would be:

1.)

```js
var e = b.ellipse(50, 50, 50, 50);
b.contentType(e, b.TEXT_TYPE);
b.typo(e, "contents", "My string."); // or alternatively use a new place function
```

as opposed to

2.)
```js
var e = b.ellipse(50, 50, 50, 50);
b.text("My string.", e);
```

I find option 2 much less confusing. I think putting a text on the page should ususally go along with a `b.text()` command. Option 1 on the other side requires a new function `b.contentType()`, some new constants and it would be hard to find in the docs if a user asks himself: "Okay, how can I place text inside a shape?"

**Comment by b-g** [comment]
and let me add something very targeted to "what I want to do" ... to our discussion:

3.)
```js
var e = b.ellipse(50, 50, 50, 50);
b.textInside("May string.", e);
```

**Comment by trych** [comment]
> `b.textInside("May string.", e);`

Sure, this would make sense, too. I just thought since the functionality is already included in `b.image()` it would make sense to include the same functionality in `b.text()` as well. But we could also split it up in `b.image()` / `b.imageInside()` and `b.text()` / `b.textInside()`.

The word "inside" does not fit too well to a path, though, but I cannot come up with a better word at the moment.

**Comment by ffd8** [comment]
I've thought about it more and more.. and like the idea of extending b.text() and b.image()... either you generate them on the fly or you simply already have a container to hold them. 

So the following would be possible:
`b.text(b.LOREM, 50, 50);` (auto builds big text box, then trims with fitting afterwards)
`b.text(b.LOREM, 50, 50, 100, 200);` (as we've had)
`b.text(b.LOREM, obj);` (places object.. we compare content_types and convert if necessary)

and for images:

`b.image(img, 50, 50);`
`b.image(img, 50, 50, 100, 100);`
`b.image(img, obj);`

But one issue comes up w/ both b.text + b.image:
`var e = b.ellipse(50, 50, 50, 50);`
`var t = b.text("blah", e);`

Which is the object now?? `t` or `e`? Since one could have called b.text with or without a var... or it doesn't matter and both will work fine?

**Comment by trych** [comment]
@ffd8 Would it be ok, if I re-assigned myself to this one as well? That's one of the things that I would like to have implemented before the 2.0 release as well.

**Comment by trych** [comment]
Placing of images in text frames and placing of text in shapes was implemented in #287. Will close this issue.
